Update attention.cpp to remove warning about preferring torch.bool type

mikekgfb opened the pull request:
Update attention.cpp to remove warning about preferring torch.bool data type 

Fixes #100469 #97532

mikekgfb commented:
Masks can be floating point (and in fact are preferred to be floating point for SDPA), so for consistency we should allow float masks for creating Nested Tensors from vanilla (dense) tensors in conjunction with a float-typed padding mask with the convention of -infinity being a mask-out

https://github.com/pytorch/pytorch/blob/03101a227f6639d5a9ad628d1dc300f9f99a8812/aten/src/ATen/native/nested/NestedTensorMath.cpp#L63
https://github.com/pytorch/pytorch/blob/03101a227f6639d5a9ad628d1dc300f9f99a8812/aten/src/ATen/native/nested/NestedTensorMath.cpp#L94

mikaylagawarecki reviewed (changes requested):
Sorry, from offline discussion with @jbschlosser, removing the warning in masked_softmax is probably ok but perhaps we should not be converting the float masks to bool masks in the NestedTensor_from_mask -- the mask seems to be getting converted from bool-->float-->bool again (perhaps the fix could be localized to[`_canonical_mask` where we are doing the conversion from bool --> float once](https://github.com/pytorch/pytorch/blob/main/torch/nn/functional.py#L5057))

jbschlosser commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -60,15 +60,23 @@ Tensor pad_tensor_to_shape(
-    Tensor sizes = mask;
+    Tensor sizes;
+    if (mask.scalar_type() != at::ScalarType::Bool){
+      sizes = mask.to(at::kBool);
```
Is my understanding here correct? If the user passes a bool mask into `TransformerEncoderLayer`, it will be converted to a float in `F._canonical_mask()`. Then it will be reconverted back to bool here.

If so, I'd prefer we avoid the conversion gymnastics. Can we just change `F._canonical_mask()` somehow to do it right in the first place?

mikekgfb replied:
Joel, a few considerations, and I'm happy to do this another way:

1 - if we say a mask can be floating point (which we say pervasively everywhere), then we shoulkd accept it here.  
2 - you are right, in this scenario for the fastpath there are two conversions. It's not so easy though...

Considerations:
* we need a very detailed set of conditions to find whether this should be float or bool.  Because when it goes to sdpa, float is the preferred way, otherwise we have a conversion in every layer.
* so, we'd have to duplicate the analysis before canonical mask -- or sink canonical mask further down.  That's feasible (at some logic complexity, like we need a "check_only" boolean, and the path will be more interesting.  But manageable
* switching to float not only avoided the conversions on the sdpa, path, it also allowed us to handle mask logic more easily.  Because masks must be folded at various points, and that uses to check wthe data type and eiother add the masks (for float masks) or do logicals (for bool masks).  So we have to add the bool mask merge logic back everywhere.
* double conversion bool-float-bool happens only once on the way to Nested Tensor creation (once for all layers), bool float conversion if a bool is passed to sdpa happens in every layer.  
* we are talking about refactoring the fastpath to take advantage. of PT2 compile (since the current Nested Tensors are not supproted by `torch.compile()`.  So just wanted to switch off annoying warnings reported by our users (and the warnings are frankly incorrect, because even if they pass in bool we convert it to float and warn the users to please pass in bool.  Ummm...)  and improve the general situation (such as allowing float masks in nested tensor creation).

This is not a perfect solution, I did feel it made things notably better.  In my mind a perfect solution is more expensive, and actually not that perfect, and not again, the refactor just screamed make it better with limited engineering.  Doing nothing feels like less than this.

mikaylagawarecki commented:
cc @i-markov

jbschlosser commented on aten/src/ATen/native/transformers/attention.cpp:
```diff
@@ -147,9 +147,6 @@ Tensor masked_softmax(
     return at::_nested_tensor_softmax_with_shape(attn_scores, query);
   }
   if (attn_mask && attn_mask->dtype() != at::kBool) {
-    TORCH_WARN(
```
I think there should be a way to avoid triggering this warning without actually removing it from the codebase.

mikekgfb replied:
Supposedly not so well, and there were particular complaints about Torchscript.  But if you can suggest a solution.  
That being said, even then, the warning is sorta incorrect, because we end up warning the user even if she passed in a bool, because we converted it for them.)

mikaylagawarecki replied:
For posterity I think this is ok because I believe the only non-transformers usecase for `masked_softmax` where this warning might be useful is MaskedTensor

jbschlosser reviewed (changes requested):
I have some concerns about the mask conversions and am hoping we can take a simpler approach here.

Disclaimer: I haven't had a ton of time to dig into this in detail.

mikekgfb commented:
Updated TransformerEncoder to keep the original input for key src padding mask and mask.  When we go to the C++ fastpath, we use the original inputs to avoid double conversions.

I suggest we keep the framework for supporting floats more efficiently by recognizing floating point masks properly for fastpath (without warnings) because for all other execution modes we actually prefer float so we shouldn't aggressively encourage users to use bool instead of float.

mikekgfb commented:
@mikaylagawarecki @jbschlosser can you please have a look to see whether this addresses Joel's comments about double conversion?

mikaylagawarecki commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -60,15 +60,23 @@ Tensor pad_tensor_to_shape(
     // N * L
-    Tensor sizes = mask;
+    Tensor sizes;
+    if (mask.scalar_type() != at::ScalarType::Bool){
```
hm, why do we still need to allow floats here and in `NestedTensor_nested_tensor_from_mask_left_aligned` if we do `src_key_padding_mask_.logical_not()` when passing the mask argument in which iiuc will return a bool mask below?

mikekgfb replied:
Removed.

mikaylagawarecki commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -61,6 +61,9 @@ Tensor pad_tensor_to_shape(
 
 Tensor NestedTensor_nested_tensor_from_mask(const Tensor& t, const Tensor& mask, bool mask_check) {
     TORCH_CHECK(mask.scalar_type() == at::ScalarType::Bool, "Expected mask to be of ScalarType Bool, but got ", mask.scalar_type(), " instead.");
```
seems like this line should be removed :)

mikekgfb replied:
That's what happens when I try to do things quick.  I agree that we're not calling this today.  So, we don't need to update the check.  (And the change does exactly that, because I added the old check back but had not taken the second check out.  )

The original change was not motivated by us calling this function with float today, but rather with the intent that "we're treating floats as legal type for masks, so let's make them legal wherever masks go".  Maybe a bit over-eager ;)
So, removing that bit and dialing it back to where it was

mikaylagawarecki commented on torch/nn/modules/transformer.py:
```diff
@@ -273,10 +278,13 @@ def forward(
             why_not_sparsity_fast_path = "src_key_padding_mask was None"
         elif (((not hasattr(self, "mask_check")) or self.mask_check)
-                and not torch._nested_tensor_from_mask_left_aligned(src, src_key_padding_mask.logical_not())):
+                and not torch._nested_tensor_from_mask_left_aligned(src, src_key_padding_mask_.logical_not())):
```
I'm sorry I didn't catch this in the previous pass but reading this again I am actually somewhat confused here, it seems like we are doing 

1) save `src_key_padding_mask` as `src_key_padding_mask_`
2)`src_key_padding_mask =  _canoncial_mask(...)` which will do bool --> float conversion
3) call `src_key_padding_mask.logical_not()` which will make the mask passed in a boolean mask again

So do we need to save `src_key_padding_mask_` at all? Similarly for `mask`

mikekgfb committed:
Update attention.cpp to remove warning about preferring torch.bool type (#103362)

Summary:
Update attention.cpp to remove warning about preferring torch.bool data type

Fixes https://github.com/pytorch/pytorch/issues/100469 #97532

Pull Request resolved: https://github.com/pytorch/pytorch/pull/103362

Reviewed By: mikaylagawarecki

Differential Revision: D46730530

Pulled By: mikekgfb

fbshipit-source-id: 046b8fef1325d6821dde9d7a2e57c3072532c011

mikaylagawarecki reviewed (approved):
Thanks!

mikekgfb commented:
@pytorchbot merge
